upup/pkg/fi/cloudup/awstasks/external_target_group_attachment.go

```diff
@@ -0,0 +1,140 @@
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kops/upup/pkg/fi/cloudup/cloudformation"
+	"k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+)
+
+//go:generate fitask -type=ExternalTargetGroupAttachment
+type ExternalTargetGroupAttachment struct {
+	Name      *string
+	Lifecycle *fi.Lifecycle
+
+	TargetGroupARN string
+
+	AutoscalingGroup *AutoscalingGroup
+}
+
+func (e *ExternalTargetGroupAttachment) Find(c *fi.Context) (*ExternalTargetGroupAttachment, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	if e.TargetGroupARN == "" {
+		return nil, fmt.Errorf("InstanceGroup did not have TargetGroupARNs set")
+	}
+
+	g, err := findAutoscalingGroup(cloud, *e.AutoscalingGroup.Name)
+	if err != nil {
+		return nil, err
+	}
+	if g == nil {
+		return nil, nil
+	}
+
+	for _, name := range g.TargetGroupARNs {
+		if aws.StringValue(name) != e.TargetGroupARN {
+			continue
+		}
+
+		actual := &ExternalTargetGroupAttachment{}
+		actual.TargetGroupARN = e.TargetGroupARN
+		actual.AutoscalingGroup = e.AutoscalingGroup
+
+		// Prevent spurious changes
+		actual.Name = e.Name // ELB attachments don't have tags
+		actual.Lifecycle = e.Lifecycle
+
+		return actual, nil
+	}
+
+	return nil, nil
+}
+
+func (e *ExternalTargetGroupAttachment) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (s *ExternalTargetGroupAttachment) CheckChanges(a, e, changes *ExternalTargetGroupAttachment) error {
+	if a == nil {
+		if e.TargetGroupARN == "" {
+			return fi.RequiredField("TargetGroupARN")
+		}
+		if e.AutoscalingGroup == nil {
+			return fi.RequiredField("AutoscalingGroup")
+		}
+	}
+	return nil
+}
+
+func (_ *ExternalTargetGroupAttachment) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *ExternalTargetGroupAttachment) error {
+	if e.TargetGroupARN == "" {
+		return fi.RequiredField("TargetGroupARN")
+	}
+
+	request := &autoscaling.AttachLoadBalancerTargetGroupsInput{}
+	request.AutoScalingGroupName = e.AutoscalingGroup.Name
+	request.TargetGroupARNs = aws.StringSlice([]string{e.TargetGroupARN})
+
+	glog.V(2).Infof("Attaching autoscaling group %q to Target Group %q", fi.StringValue(e.AutoscalingGroup.Name), e.TargetGroupARN)
+	_, err := t.Cloud.Autoscaling().AttachLoadBalancerTargetGroups(request)
+	if err != nil {
+		return fmt.Errorf("error attaching autoscaling group to ELB: %v", err)
+	}
+
+	return nil
+}
+
+type terraformExternalTargetGroupAttachment struct {
+	TargetGroupARN   *terraform.Literal `json:"alb_target_group_arn,omitempty"`
+	AutoscalingGroup *terraform.Literal `json:"autoscaling_group_name,omitempty"`
+}
+
+func (_ *ExternalTargetGroupAttachment) RenderTerraform(t *terraform.TerraformTarget, a, e, changes *ExternalTargetGroupAttachment) error {
+	tf := &terraformExternalTargetGroupAttachment{
+		TargetGroupARN:   terraform.LiteralFromStringValue(e.TargetGroupARN),
+		AutoscalingGroup: e.AutoscalingGroup.TerraformLink(),
+	}
+
+	return t.RenderResource("aws_autoscaling_attachment", *e.Name, tf)
```

Comment:
Sorry if I was a bit unclear about this on my previous comment. `*e.Name` won't actually work here as is because it contains `:` in it since the full ARN contains those. The way you had it before was good with `e.name()` but you you need to attach something else to it such as the `e.AutoscalingGroup.Name` to make it unique. Or you could always just sanitize `*e.Name` and remove the `:` from it.

Reply:
I noticed slashes were already being sanitized to dashes, I assumed (wrongly) that it'd sanitize the colons as well. I'll revert it a bit.